core/server/api/v2/oembed.js

```diff
@@ -0,0 +1,102 @@
+const common = require('../../lib/common');
+const {extract, hasProvider} = require('oembed-parser');
+const Promise = require('bluebird');
+const request = require('../../lib/request');
+const cheerio = require('cheerio');
+
+const findUrlWithProvider = (url) => {
+    let provider;
+
+    // build up a list of URL variations to test against because the oembed
+    // providers list is not always up to date with scheme or www vs non-www
+    let baseUrl = url.replace(/^\/\/|^https?:\/\/(?:www\.)?/, '');
+    let testUrls = [
+        `http://${baseUrl}`,
+        `https://${baseUrl}`,
+        `http://www.${baseUrl}`,
+        `https://www.${baseUrl}`
+    ];
+
+    for (let testUrl of testUrls) {
+        provider = hasProvider(testUrl);
+        if (provider) {
+            url = testUrl;
+            break;
+        }
+    }
+
+    return {url, provider};
+};
+
+const getOembedUrlFromHTML = (html) => {
+    return cheerio('link[type="application/json+oembed"]', html).attr('href');
+};
+
+module.exports = {
+    docName: 'oembed',
+
+    read: {
+        permissions: false,
+        data: [
+            'url'
+        ],
+        options: [],
+        query({data}) {
+            let {url} = data;
+
+            if (!url || !url.trim()) {
```

Comment:
IMO this is validation stage. any reason you have not used this stage?

(furthermore: !url can be achieved with requried:true)

Reply:
But in case we put this validation into earlier stage we'll have less specific error returned by the server (atm it's `message: common.i18n.t('errors.api.oembed.noUrlProvided')`) and it will become just a generic 400 right? But maybe that message is not important at all, then def could be moved

Reply:
Can be generic IMO. Can we track that in #10060? @rishabhgrg